### PR TITLE
Deprecate config item `max-memory` and add items `server-memory-quota` and `memory-usage-alarm-ratio` (#4977)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #4977 to release-4.0

---

### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Deprecate config item `max-memory` and add items `server-memory-quota` and `memory-usage-alarm-ratio` (only in dev)

`max-memory` has been deprecated and use `server-memory-quota` instead now. (only in dev)
`memory-usage-alarm-ratio` is introduced in this pr https://github.com/pingcap/tidb/pull/18858.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions) (need a special version)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s): https://github.com/pingcap/docs-cn/pull/4706

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [x] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Review by TomShawn — CHANGES_REQUESTED** [review]
@wshwsh12

**Review by XuHuaiyu — COMMENTED** [review]
LGTM

File: tidb-configuration-file.md [hunk 1]
```diff
@@ -283,6 +283,25 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
 + 默认值：0
 + 这个配置只有在 `prepared-plan-cache.enabled` 为 `true` 的情况才会生效。当 LRU 的 size 大于 `prepared-plan-cache.capacity` 时，也会剔除 LRU 中的元素。
 
+### `server-memory-quota`
+
+> **警告：**
+>
+> `server-memory-quota` 目前为实验性特性，不建议在生产环境中使用。
+
++ tidb-server 实例内存的使用限制，单位为字节。<!-- 从 TiDB v5.0 起 -->该配置项完全取代原有的 [`max-memory`](https://docs.pingcap.com/zh/tidb/stable/tidb-configuration-file#max-memory)。
```
Comment (TomShawn) [inline tidb-configuration-file.md:292]: 是否需要在 `max-memory` 描述部分说明一下，推荐使用哪一个配置项？并说明 `max-memory` 会在未来版本中完全被 `server-memory-quota` 取代？
```suggestion
+ tidb-server 实例内存的使用限制，单位为字节。在未来版本中，该配置项将完全取代原有的 [`max-memory`](/tidb-configuration-file#max-memory)。
```

File: tidb-configuration-file.md [hunk 2]
```diff
@@ -283,6 +283,24 @@ TiDB 配置文件比命令行参数支持更多的选项。你可以在 [config/
...
++ tidb-server 实例内存的使用限制，单位为字节。
++ 默认值：0
++ 默认值为 0 表示无内存限制。
+
+### `memory-usage-alarm-ratio`
+
++ tidb-server 实例内存使用占总内存的比例超过一定阈值时会报警。该配置项的有效范围为 `0` 到 `1`。如果配置该选项为 `0` 或 `1`，则表示关闭内存阈值报警功能。
++ 默认值：0.8
++ 当内存阈值报警功能开启时，如果配置项 [`server-memory-quota`](/tidb-configuration-file.md#server-memory-quota) 未设置，则内存报警阈值为 `memory-usage-alarm-ratio * 系统内存大小`；如果 `server-memory-quota` 被设置且大于 0，则内存报警阈值为 `memory-usage-alarm-ratio * server-memory-quota`。
++ 当 TiDB 检测到 tidb-server 的内存使用超过了阈值，则会认为存在内存溢出的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
```
Comment (TomShawn) [inline tidb-configuration-file.md:301]:
```suggestion
+ 当 TiDB 检测到 tidb-server 实例的内存使用超过了阈值，则会认为存在内存溢出的风险，会将当前正在执行的所有 SQL 语句中内存使用最高的 10 条语句和运行时间最长的 10 条语句以及 heap profile 记录到目录 [`tmp-storage-path/record`](/tidb-configuration-file.md#tmp-storage-path) 中，并输出一条包含关键字 `tidb-server has the risk of OOM` 的日志。
```

**Review by TomShawn — APPROVED** [review]
LGTM

**Pull request merged**